### DISCUSSION TITLE
How does Noice render its cursor in cmdline?

### DISCUSSION THREAD

**Discussion by savq** [folke/noice.nvim#596] [Q&A]
A user has opened an issue on my colorscheme's issue tracker regarding the fact that the cursor in noice's cmdline becomes invisible. https://github.com/savq/melange-nvim/issues/67. I've been trying to look at what could be causing the issue, but it still isn't clear to me how noice renders cursors.

I looked at the issue tracker here and was able to find some commends by Folke on the topic:

In https://github.com/folke/noice.nvim/issues/552#issuecomment-1652159334:
> not possible to fix, since what you see in the noice cmdline is not an actual cursor.

In https://github.com/folke/noice.nvim/issues/386#issuecomment-1457686448:
> NoiceCursor links to Cursor. Your colorscheme is responsible to assign a proper highlight to that group. The actual colors used to remnder the actual cursor are chosen by your terminal, not your colorscheme.

The latter comment is specially confusing, since Neovim already assigns a highlight group to `Cursor` (see `:hi clear | hi Cursor` and [:h E420](https://neovim.io/doc/user/syntax.html#E420)). It seems strange that noice doesn't work with the default highlight group.

My question here is not only "why is noice reinventing the wheel?", but rather "why is noice reinventing the wheel so poorly that Neovim defaults don't work?"

**Comment by folke** [comment]
This is **not** a Noice issue. The actual cursor being rendered in Neovim is rendered by your terminal, **not** by the `Cursor` highlight group. I know this is confusing, but that is how Neovim/terminals work.

A colorscheme is supposed to set the `Cursor` highlight group to something that makes sense so that other plugins (like Noice) can render something resembling a cursor.

There is no reinventing the wheel here. Noice uses `ui_attach` to render the cmdline ui. That does not support the actual cursor position, so the only way to show the cursor is by rendering a fake cursor using the `Cursor` highlight group.

The same approach is used in a lot of plugins like Hop, Flash, ...
